### PR TITLE
Add JTAG driver for libgpiod bitbanging

### PR DESCRIPTION
**Pull request by niklasekstrom**
A driver is added that allows bitbanging JTAG on a Linux host with GPIO pins, such as a raspberry pi.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -11,6 +11,7 @@ else()
 	option(ENABLE_UDEV "use udev to search JTAG adapter from /dev/xx" ON)
 endif()
 option(ENABLE_CMSISDAP "enable cmsis DAP interface (requires hidapi)" ON)
+option(ENABLE_LIBGPIOD "enable libgpiod bitbang driver (requires libgpiod)" ON)
```
Comment (trabucayre) [inline CMakeLists.txt:14]: Maybe here or below a test about OS. I'm not familiar with macOS and windows and not sure `/dev/gpiochipxx` is something allowed.

File: src/cable.hpp [hunk 2]
```diff
@@ -69,6 +70,7 @@ static std::map <std::string, cable_t> cable_list = {
 	{"usb-blaster",        {MODE_USBBLASTER,   {0x09Fb, 0x6001, 0,           0,    0,    0,    0,    0}}},
 	{"usb-blasterII",      {MODE_USBBLASTER,   {0x09Fb, 0x6810, 0,           0,    0,    0,    0,    0}}},
 	{"xvc-client",         {MODE_XVC_CLIENT,   {}}},
+	{"libgpiod",           {MODE_LIBGPIOD_BITBANG, {}}},
```
Comment (trabucayre) [inline src/cable.hpp:73]: `#ifdef ENABLE_LIBGPIOD` to avoid to display this cable if support is disabled?

File: src/jtag.cpp [hunk 3]
```diff
@@ -130,6 +133,13 @@ void Jtag::init_internal(cable_t &cable, const string &dev, const string &serial
 #else
 		std::cerr << "Jtag: support for xvc-client was not enabled at compile time" << std::endl;
 		throw std::exception();
+#endif
+#ifdef ENABLE_LIBGPIOD
+	case MODE_LIBGPIOD_BITBANG:
+		if (pin_conf == NULL)
+			throw std::exception();
```
Comment (trabucayre) [inline src/jtag.cpp:140]: An error message to explain why there is a failure?
And as for **xvc** and **cmsis** an `#else` to display a message when user try to use this driver if it's disabled.

File: src/libgpiodJtagBitbang.cpp [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+		for (int j = i + 1; j < sizeof(pins) / sizeof(pins[0]); j++) {
+			if (pins[i] == pins[j]) {
+				display("Two or more pins are assigned to the same pin number %d\n", pins[i]);
+				throw std::runtime_error("Two or more pins are assigned to the same pin number\n");
+			}
+		}
+	}
+
+	_chip = gpiod_chip_open_by_name(chip_dev.substr(5).c_str());
```
Comment (trabucayre) [inline src/libgpiodJtagBitbang.cpp:72]: `gpio_chip_open` takes a `/dev/gpiochipxx` it's avoid to cut `/dev/`

File: src/libgpiodJtagBitbang.cpp [hunk 5]
```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * Copyright (C) 2020 Gwenhael Goavec-Merou <gwenhael.goavec-merou@trabucayre.com>
```
Comment (trabucayre) [inline src/libgpiodJtagBitbang.cpp:3]: Wrong copyright :)

File: src/libgpiodJtagBitbang.hpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * Copyright (C) 2020 Gwenhael Goavec-Merou <gwenhael.goavec-merou@trabucayre.com>
```
Comment (trabucayre) [inline src/libgpiodJtagBitbang.hpp:3]: ditto

File: src/libgpiodJtagBitbang.cpp (hunk 5 above)
Comment (niklasekstrom) [reply src/libgpiodJtagBitbang.cpp:3]: In what way is the copyright wrong? I based the code on the file ftdiJtagBitbang.cpp, and then added a line under your name that says that the libgpiod code was added by me.

File: src/jtag.cpp (hunk 3 above)
Comment (niklasekstrom) [reply src/jtag.cpp:140]: That code is copy-pasted from the MODE_FTDI_BITBANG case above in the switch statement.

It looks like the `if (pin_conf == NULL)` check is not needed as pin_conf will always be non-NULL, seen by looking at the call site, so I'll just remove the check.

I'll add the `#else` message you suggested.

File: src/libgpiodJtagBitbang.cpp (hunk 4 above)
Comment (niklasekstrom) [reply src/libgpiodJtagBitbang.cpp:72]: Done.

File: src/cable.hpp (hunk 2 above)
Comment (niklasekstrom) [reply src/cable.hpp:73]: Will do.

File: src/libgpiodJtagBitbang.cpp (hunk 5 above)
Comment (trabucayre) [reply src/libgpiodJtagBitbang.cpp:3]: I'm not really aware about copyright rule (I must must learn more about).
This class is based on another but with modifications so it seems made sense to add you in copyright are adding a note like "based on xxx". You have done the job to adapts code for libgpiod.
Or simply adding a second copyright line :)

**Comment by trabucayre** [comment]
LGTM.

**Comment by trabucayre** [comment]
Applied to master
thanks!

**Pull request merged**